File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxDepth(TreeNode* root) {
        

        function<int(TreeNode*)> go = [&](TreeNode* root) {
            if (root == nullptr) return 0;
            return 1 + max(root->left ? go(root->left) : 0, root->right ? go(root->right) : 0);
        };

        return go(root);
    }
};

File: Data Structures & Algorithms/invert-a-binary-tree/submission-0.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func invertTree(root *TreeNode) *TreeNode {
    if root == nil {
        return nil
    }

    var newRoot TreeNode
    newRoot.Val = root.Val

    if root.Left != nil {
        newRoot.Right = invertTree(root.Left)
    }


    if root.Right != nil {
        newRoot.Left = invertTree(root.Right)
    }


    return &newRoot
}
